### ISSUE TITLE
Jaxtyping argument consistency checks are enabled at test time but not runtime

### ISSUE THREAD

**Issue by rusheb** [neelnanda-io/TransformerLens#190]
## Problem statement

Jaxtyping [multi-argument consistency checks](https://github.com/google/jaxtyping/blob/main/API.md#runtime-type-checking) are enabled at test time but not at runtime, resulting in different behaviour. This prevents us from testing all runtime behaviour in unit tests.

For example, consider calling the [FactoredMatrix](https://github.com/neelnanda-io/TransformerLens/blob/main/transformer_lens/FactoredMatrix.py) constructor with a mismatched inner dimension:
```python
A = torch.ones((2, 3, 4))
B = torch.ones((2, 3, 5))
FactoredMatrix(A, B)
```

At runtime, this will lead to an AssertionError due to an assertion made in the constructor:
```python
AssertionError: Factored matrix must match on inner dimension, shapes were a: torch.Size([2, 3, 4]), b:torch.Size([2, 3, 5])
```

However, at test-time, when running [`test_factored_matrix_inner_mismatch`](https://github.com/rusheb/TransformerLens/blob/f11c3792d9a99b1f94689279b1756fe4fab9b93d/transformer_lens/tests/factored_matrix/test_constructor.py#L63) (note: this test is not yet merged into main), we see a different error:
```python
TypeError: type of argument "B" must be jaxtyping.Float[Tensor, '... mdim rdim']; got torch.Tensor
```

The TypeError is triggered by jaxtyping/typeguard runtime typechecking on the constructor arguments. This error is thrown before we reach the assertion, meaning the AssertionError is not thrown. 

## Mitigation
It's not clear what the ideal solution is. I think these are the options:

* **Enable multi-argument consistency checks always.** This could be achieved using `install_import_hook` or the `@jaxtyping` and `@typechecked` annotations. However, I'm not sure what unintended sied-effects this would cause. On another note, the error message for the TypeError is actually misleading/confusing. The AssertionError has a much more clear and descriptive message. 

* **Disable multi-argument consistency checks always.** I'm not sure how we would disable it at test time. We need to get to the bottom of why it is occurring. I suspect it is related to calling `hook = install_import_hook("transformer_lens", ("typeguard", "typechecked"))` from the test files e.g. `test_stop_at_layer.py`. However in each of these files, the call to `hook.uninstall()` at the end of the file should prevent this from affecting other files.

* **Suppress runtime typechecking in specific tests only.** Typeguard 3 supports [typecheck suppression](https://typeguard.readthedocs.io/en/latest/userguide.html#suppressing-type-checks) However typeguard 3 is still a release candidate; we would need to upgrade.

---

**Comment by dkamm** [comment]
Hey @rusheb - thanks for writing this up!  You're correct that `install_import_hook` is what's enabling the typechecking in the tests.

Ah man...

So the error message issue isn't just for multi-arg consistency checks- it's a general issue.  See https://github.com/google/jaxtyping/issues/6.  Patrick got back to me and he's waiting on beartyping to add new features for a proper fix.

Torchtyping actually didn't have this error message problem because of it patches typeguard.  I'm wondering if we could try modifying that code and adding it here.  However, Patrick warned me that this was a crazy solution 😆 

So I think the options are

1. Go back to torchtyping
2. Enable typechecking everywhere (even if patching typeguard doesn't work)
3. Disable typechecking until https://github.com/google/jaxtyping/issues/6 gets fixed 

If patching typeguard doesn't work, I think we should go with option 3.  My second choice would be option 1.

The error message issue is confusing enough that I don't think we should do option 2 (which sadly defeats the purpose of having a typechecker).  I think it would take a while to "get used" to and would be especially confusing to new users of the library.

**Comment by dkamm** [comment]
I can take a stab at patching typeguard btw

**Comment by rusheb** [comment]
Thanks so much for getting back! Your recommendations sound reasonable to me - I'm not too sure of the best option. Maybe best to wait to hear back from a maintainer?

**Comment by jbloomAus** [comment]
I'm sorry; this must have fallen through the cracks. My takes are: 
- Almost certainly doesn't make sense to go back to torchtyping (unless this becomes more painful).
- @rusheb Typeguard 3 appears to have been released? Maybe we just upgrade and suppress in specific tests until...
- When https://github.com/google/jaxtyping/issues/6 is closed so we can have informative error messages and enable multi-argument consistency checks always.

If this sounds like a good plan to you @rusheb and @dkamm I'll make a checklist and we can go from there.

**Comment by neelnanda-io** [comment]
@jbloomAus My understanding is that the JaxTyping issue has been open for almost a year, and so probably won't be closed any time soon? (The beartypign thing discussed was a more recent update, but was still a month ago). So we should pursue a solution that doesn't depend on it.

I don't otherwise have strong takes, though I don't see what's that bad about going back to torchtyping? (IIRC the main problem was bad docstrings?). Investigating typeguard 3 also seems reasonable to me.

But yeah, idk, this is not my area of expertise, and I'll happily defer to other people on this.

**Comment by dkamm** [comment]
Hey guys-

Sorry I dropped the ball on this.

TLDR: I think we can get something "good enough" by writing our own monkey patch, at least until that `beartyping` change is implemented and `jaxtyping` is updated to use it.  @jbloomAus I can send a PR sometime this week if that sounds good.

I monkey patched the `qualified_name` method to make it return dtype and shape information.  It works ok for single argument checks.  The error message for multi-argument consistency checks leaves much to be desired, but requires more monkey patching to fix and I haven't investigated that yet.  It might be better to just disable multi-argument consistency checks always if this ends up being too brittle.

Here's my script

```python
import typeguard
import string
from jaxtyping import Float, jaxtyped
import torch
from typing import Any, Union


def patch_typeguard():

    orig_qualified_name = typeguard._utils.qualified_name

    def qualified_name(obj: Any, *args, add_class_prefix: bool = False) -> str:
        if isinstance(obj, torch.Tensor):
           
            dtype = str(obj.dtype)
            if torch.is_floating_point(obj):
                dtype = 'Float'
            elif dtype.startswith('torch.int'):
                dtype = 'Int'
            elif dtype.startswith('torch.bool'):
                dtype = 'Bool'

            # Can display the actual dims or just letters
            dims = " ".join(str(d) for d in obj.shape)
            #dims = " ".join([c for c in string.ascii_lowercase[:obj.dim()]])

            return f"jaxtyping.{dtype}[torch.Tensor, '{dims}']"

        return orig_qualified_name(obj, *args, add_class_prefix=add_class_prefix)

    # TODO: need to patch all imports of `qualified_name`
    typeguard._utils.qualified_name = qualified_name
    typeguard._functions.qualified_name = qualified_name


patch_typeguard()


@jaxtyped
@typeguard.typechecked
def foo(x: Float[torch.Tensor, "a b"]):
    return x.mean(dim=-1)

@jaxtyped
@typeguard.typechecked
def bar(x: Union[float, Float[torch.Tensor, "a"]]):
    return x

@jaxtyped
@typeguard.typechecked
def baz(x: Float[torch.Tensor, "a b"], y: Float[torch.Tensor, "b c"]):
    return x @ y


try:
    x = torch.tensor([False, True, True])
    print(f"calling foo({x})")
    foo(x)
except typeguard.TypeCheckError as err:
    print("got error: ", err)

try:
    x = torch.tensor([[1, 2, 3], [4, 5, 6]])
    print(f"calling bar({x})")
    bar(x)
except typeguard.TypeCheckError as err:
    print("got error: ", err)


try:
    x = torch.randn((1,2))
    y = torch.randn((3,2))
    print(f"calling baz({x},{y})")
    baz(x,y)
except typeguard.TypeCheckError as err:
    print("got error: ", err)
```

and terminal output:

```
calling foo(tensor([False,  True,  True]))
got error:  argument "x" (jaxtyping.Bool[torch.Tensor, '3']) is not an instance of jaxtyping.Float[Tensor, 'a b']
calling bar(tensor([[1, 2, 3],
        [4, 5, 6]]))
got error:  argument "x" (jaxtyping.Int[torch.Tensor, '2 3']) did not match any element in the union:
  float: is neither float or int
  jaxtyping.Float[Tensor, 'a']: is not an instance of jaxtyping.Float[Tensor, 'a']
calling baz(tensor([[ 0.6991, -0.8888]]),tensor([[-1.0327,  0.0934],
        [ 1.1167, -0.2152],
        [ 0.9708,  0.0119]]))
got error:  argument "y" (jaxtyping.Float[torch.Tensor, '3 2']) is not an instance of jaxtyping.Float[Tensor, 'b c']
```

I agree that it doesn't make sense to go back to `torchtyping`.  `torchtyping` also used monkey patching (see https://github.com/patrick-kidger/torchtyping/blob/master/torchtyping/typechecker.py#L263) and it looks like this monkey patch doesn't work with `typeguard` 3 anyway

**Comment by jbloomAus** [comment]
@neelnanda-io Oh I see what you mean with the `beartyping` issue. That's a shame. In terms of going back to `torchtyping`, the principle that comes to mind here is more SE related for me. It's something like "have a high bar for changes that affect many lines of the code" (ie: unless something is a big issue, be reluctant to make that kind of change) and two, don't undo recent changes. They're rough guidelines but given the motivation on this issue is that something is slightly annoying to test and not an actual bug, the principles win out in my mind.

Hmmm. I think the monkey patch could be good but agree if it turns out too brittle we should abandon it. If we get more issues/feedback on this or it comes up again as we continue to expand test coverage, then we can either decide to improve the monkey patch or to disable multi-argument consistency checks.

I appreciate the support! Thanks for helping with this!

**Comment by alan-cooney** [comment]
Closing as stale - feel free to return to the PR and re-open if you want however
